### DISCUSSION TITLE
Save changed username in database and use that

### DISCUSSION THREAD

**Discussion by karaka200** [symfony/symfony#45984] [Q&A]
At the time of registration, I change the users username(mobile number):

```
public function setUsername(string $username): self
{
	$array = [
		'0' => '1',
		'1' => '4',
		'2' => '7',
		'3' => '0',
		'4' => '3',
		'5' => '6',
		'6' => '9',
		'7' => '2',
		'8' => '5',
		'9' => '8'
	];
	$invert = strrev($username);
	$this->username = strtr($invert, $array);
	return $this;
}
```

but when users want to login, display this error:

`Invalid credentials.`

How can I change the username first when login?

**Comment by MichaelBrauner** [comment]
Ok, you set the username on registration - than they want to login and this does not work even if they type in the correct username - password combination? Have I understood that correctly?

**Reply by karaka200** [reply]
> Ok, you set the username on registration - than they want to login and this does not work even if they type in the correct username - password combination? Have I understood that correctly?

@MichaelBrauner  Yes, that's right

**Comment by shmolf** [comment]
Some questions I have:
- Why reverse the input username?
- Why translate the username to a different sequence of numbers?

If the purpose is to obfuscate their username, then why not store their username as a sha hash?

---

As for why they can't login:

W/o looking at the code, it's hard to discern, but the controller that receives the login username would need to _transform_ this number, just as being done in this `setUsername` method.

I would recommend a separate class who's responsible for translating the username into it's final form. This way, a single piece of code can be used by the Entity/Controller/Service.
```php
class UsernameTransformer {
  public function obfuscate(string $username): string
  {
    return hash('sha256', $username);
  }
}
```

Then, at your callsite:
```php
public function setUsername(string $username): self
{
  $this->username = (new UsernameTransformer()).obfuscate($username);
  return $this;
}
```

**Reply by karaka200** [reply]
Hash is difficult to decode and I only do so because if a hacker accesses the database hacker can not get the mobile number of people.
Your code is good but how can I decode `$username` hash? 
for example my number ( `$username` ) is `56745598745`, hash code is `90c850fc02c5b5286b14a332895f7055c6778b26c55d5d28038297e5c0964f57` but I can't decode this hash for use in my site.

**Reply by shmolf** [reply]
You cannot decode the hash, which is the point.

If a hacker can get to your database, they can more likely access your code as well, in which case they can see the obfuscation technique.

If you want to prevent the hacker's access, then I recommend strong security policies, including complex passwords that aren't shared.
If the concern is more likely a possibly bad agent within the company/organization, I recommend a separate ~~table~~ schema for user accounts, who's access is more restricted.

**Reply by MichaelBrauner** [reply]
Like @shmolf  said - a sha256 can not be decoded so easy - but you can check against it inside your custom UserProvider (`hash_equals`). 

The only problem I see then is that you can't show usernames inside your application. 
Since using an obfuscating strategy is not improving security I would not change usernames at all.

**Reply by shmolf** [reply]
@karaka200 Please submit your own answer, if you opted for a new solution. An update would also be appreciated.

**Comment by stof** [comment]
if your database contains the obfuscated usernames, you need to implement a user provider that can query based on the obfuscated name and not the user-provided value (which would find nothing, or find a different user).

however, I agree with @shmolf that this obfuscation is not actually improving security.
